test/e2e/framework/conformance/conformance.go

```diff
@@ -0,0 +1,946 @@
+	//
+	// May modify and return the input object.
+	GetUpdateStatus(object *unstructured.Unstructured) *unstructured.Unstructured
+
+	// GetPatchSpec describes how to generate patches.
+	//
+	// Each patch is applied to the initial object by itself, without the other patches.
+	// An empty slice is valid and disables testing of patching. This may not be sufficient
+	// for full conformance testing of the resource.
+	//
+	// If content verification is enabled, then this must cause the same change as GetUpdateSpec
+	// because verification of the patch result uses the GetUpdateSpec result as reference.
+	GetPatchSpec() []Patch
+
+	// GetPatchStatus is like GetPatchSpec for the status.
+	//
+	// The initial object with the updated spec gets patched,
+	// so the result must match the result of GetUpdateStatus
+	// applied to GetInitialObject if content verification is
+	// enabled.
+	GetPatchStatus(object *unstructured.Unstructured) []Patch
+}
+
+// Patch contains the parameters for a Patch API call.
+//
+// The data must match an existing object.
+//
+// There's no retry loop because of conflicts, so the patch should not include
+// a check of the ResourceVersion. Checking the UID in the patch is encouraged to prevent
+// patching a replaced resource.
+type Patch struct {
+	GetData func(object *unstructured.Unstructured) []byte
+	Type    types.PatchType
+}
+
+// ResourceTestcase provides test data for testing operations for a resource.
+// Test data is based on the native Go type of the resource.
+// The template parameter must be a pointer to the native Go type.
+//
+// The data is used like this:
+// - create InitialSpec -> update with UpdateSpec -> update status with UpdateStatus
+// - create InitialSpec -> apply StrategicMergePatchSpec and compare against UpdateSpec -> apply StrategicMergePatchStatus and compare against UpdateStatus
+type ResourceTestcase[T runtime.Object] struct {
+	// GroupResourceVersion identifies the API group, version and resource (plural form, lower case)
+	// within that API which is to be tested.
+	GVR schema.GroupVersionResource
+
+	// Namespaced must be true if the resource must be created in a
+	// namespace, false if it is cluster-scoped. Leaving it unset is
+	// an error.
+	//
+	// Namespaced resources get created in the test namespace.
+	//
+	// The name of cluster-scoped resources gets extended with
+	// `-<test namespace name>` to make it unique.
+	Namespaced *bool
+
+	// ContentVerificationEnabled defines whether the content of objects returned by
+	// the apiserver gets compared against the content that was sent.
+	//
+	// If enabled, all field values that were sent must also be included
+	// in the returned object. Additional fields and list or map entries
+	// may get added (for example, because of defaulting or mutating
+	// admission).
+	//
+	// This should not be enabled in conformance tests because admission
+	// is allowed to modify what is being stored.
+	ContentVerificationEnabled bool
```

Comment:
Seems like there's room to validate this struct, e.g. checking that if this bool is set then StrategicMergePatchSpec needs to be set if UpdateSpec is set?

I like what we're doing here overall, but it seems like if you don't copy-paste from another test, it might be easy to use incorrectly, even though it's still easier than writing these from scratch.